### PR TITLE
DENG-807 Added backfill complete cli command

### PR DESCRIPTION
**Pull request by wwyc**
[DENG-807](https://mozilla-hub.atlassian.net/browse/DENG-807):

* Added backfill complete cli command to complete backfill entries from backfill.yaml file:
1) clone production data
2) copy backfill data over to production data
3) delete backfill staging table

[DENG-807]: https://mozilla-hub.atlassian.net/browse/DENG-807?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: bigquery_etl/cli/backfill.py [hunk 1]
```diff
@@ -368,3 +374,142 @@ def process(ctx, qualified_table_name, sql_dir, project_id, dry_run):
...
+    except NotFound:
+        # clone production table
+        clone_query_arguments = ["cp"]
+        clone_query_arguments.append("--clone")
+        clone_query_arguments.append("-n")
+        clone_query_arguments.append(f"{project}:{dataset}.{table}")
+        clone_query_arguments.append(
+            f"{BACKFILL_DESTINATION_PROJECT}:{BACKFILL_DESTINATION_DATASET}.{cloned_table_id}"
+        )
+        subprocess.check_call(["bq"] + clone_query_arguments)
```
Comment (ANich) [inline bigquery_etl/cli/backfill.py:500]: Discussed offline - client copy job: 
`client.copy_table` 

`copy_job_config = bigquery.CopyJobConfig(operation_type=bigquery.job.copy_.OperationType.COPY, destination_expiration_time='...')`

File: bigquery_etl/cli/backfill.py [hunk 2]
```diff
@@ -368,3 +374,142 @@ def process(ctx, qualified_table_name, sql_dir, project_id, dry_run):
...
+    for backfill_date in dates:
+        partition = backfill_date.strftime("%Y%m%d")
+        if backfill_date not in entry_to_complete.excluded_dates:
+            cp_query_arguments = ["cp"]
+            cp_query_arguments.append("-f")
+            cp_query_arguments.append(
+                f"{BACKFILL_DESTINATION_PROJECT}:{BACKFILL_DESTINATION_DATASET}.{backfill_table_id}${partition}"
+            )
+            cp_query_arguments.append(f"{project}:{dataset}.{table}${partition}")
+            subprocess.check_call(["bq"] + cp_query_arguments)
```
Comment (ANich) [inline bigquery_etl/cli/backfill.py:461]: Discussed offline, same comment as below - `client.copy_table`

File: bigquery_etl/cli/backfill.py [hunk 3]
```diff
@@ -368,3 +374,142 @@ def process(ctx, qualified_table_name, sql_dir, project_id, dry_run):
...
+    _clone_table(qualified_table_name, entry_to_complete.entry_date, client)
+
+    # copy backfill data to production data
+    start_date = entry_to_complete.start_date
+    end_date = entry_to_complete.end_date
+    dates = [start_date + timedelta(i) for i in range((end_date - start_date).days + 1)]
+
+    backfill_table_id = f"{table}_{entry_to_complete.entry_date}".replace("-", "_")
+
+    for backfill_date in dates:
```
Comment (ANich) [inline bigquery_etl/cli/backfill.py:452]:
```suggestion
 
    # Replace partitions that have been backfilled
    for backfill_date in dates:
```

File: bigquery_etl/cli/backfill.py [hunk 4]
```diff
@@ -368,3 +373,136 @@ def process(ctx, qualified_table_name, sql_dir, project_id, dry_run):
...
+            """
+        )
+        sys.exit(1)
+
+    project, dataset, table = qualified_table_name_matching(qualified_table_name)
+
+    # clone production table
+    cloned_table_id = f"{table}_backup_{entry_to_complete.entry_date}".replace("-", "_")
+    cloned_table_full_name = f"{BACKFILL_DESTINATION_PROJECT}.{BACKFILL_DESTINATION_DATASET}.{cloned_table_id}"
+    _copy_table(qualified_table_name, cloned_table_full_name, client, True)
```
Comment (ANich) [inline bigquery_etl/cli/backfill.py:445]:
```suggestion
    _copy_table(qualified_table_name, cloned_table_full_name, client, clone=True)
```

File: bigquery_etl/cli/backfill.py [hunk 5]
```diff
@@ -368,3 +373,136 @@ def process(ctx, qualified_table_name, sql_dir, project_id, dry_run):
...
+    end_date = entry_to_complete.end_date
+    dates = [start_date + timedelta(i) for i in range((end_date - start_date).days + 1)]
+
+    # replace partitions in production table that have been backfilled
+    for backfill_date in dates:
+        partition = backfill_date.strftime("%Y%m%d")
+        if backfill_date not in entry_to_complete.excluded_dates:
+            production_table = f"{qualified_table_name}${partition}"
+            backfill_table = f"{backfill_staging_qualified_table_name}${partition}"
+            _copy_table(backfill_table, production_table, client)
```
Comment (ANich) [inline bigquery_etl/cli/backfill.py:458]: nit:
```suggestion
    for backfill_date in dates:
        if backfill_date in entry_to_complete.excluded_dates:
            click.echo(f"Skipping excluded date: {backfill_date})
            continue
            
        partition = backfill_date.strftime("%Y%m%d")
        production_table = f"{qualified_table_name}${partition}"
        backfill_table = f"{backfill_staging_qualified_table_name}${partition}"
        _copy_table(backfill_table, production_table, client)
```

File: bigquery_etl/cli/backfill.py [hunk 6]
```diff
@@ -368,3 +373,136 @@ def process(ctx, qualified_table_name, sql_dir, project_id, dry_run):
...
+
+    try:
+        client.copy_table(
+            source_table,
+            destination_table,
+            job_config=copy_config,
+        ).result()
+    except Exception as e:
+        print(f"Unable to clone table {source_table} with error: {e}")
+        sys.exit(1)
```
Comment (ANich) [inline bigquery_etl/cli/backfill.py:504]: What's the usual Exception here? Might be worth catching directly

**Review by ANich — APPROVED** [review]
LGTM! Left a few nits/questions

File: bigquery_etl/cli/backfill.py (hunk 6 above)
Comment (wwyc) [reply bigquery_etl/cli/backfill.py]: The exception has been expanded to NotFound and Conflict.

**Pull request merged**